modules/frontend/frontend.go

```diff
@@ -0,0 +1,227 @@
+				Name:      "query_frontend_split_queries_total",
+				Help:      "Total number of underlying query requests after sharding",
+			}, []string{"user"}),
+		}
+	})
+}
+
+type shardQuery struct {
+	next     Handler
+	shardNum int
+	logger   log.Logger
+	// Metrics.
+	splitByCounter *prometheus.CounterVec
+}
+
+// Do implements Handler
+func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
+	level.Info(s.logger).Log("msg", "shardQuery called")
+
+	userID, err := user.ExtractOrgID(r.Context())
+	if err != nil {
+		return nil, err
+	}
+
+	// create sharded queries
+	boundaryBytes := make([][]byte, s.shardNum+1)
+	for i := 0; i < s.shardNum+1; i ++ {
+		boundaryBytes[i] = make([]byte, 0)
+	}
+	const MaxUint = ^uint64(0)
+	const MaxInt = int64(MaxUint >> 1)
+	for i := 0; i < s.shardNum; i++ {
+		binary.PutVarint(boundaryBytes[i], MaxInt*(int64(i))/int64(s.shardNum))
+		binary.PutVarint(boundaryBytes[i], 0)
+	}
+	binary.PutVarint(boundaryBytes[s.shardNum], MaxInt)
+	binary.PutVarint(boundaryBytes[s.shardNum], MaxInt)
+
+	reqs := make([]*http.Request, s.shardNum)
+	for i := 0; i < s.shardNum; i++ {
+		reqs[i] = r
+		reqs[i].URL.Query().Add("blockStart", string(boundaryBytes[i]))
+		reqs[i].URL.Query().Add("blockEnd", string(boundaryBytes[i+1]))
+	}
+	s.splitByCounter.WithLabelValues(userID).Add(float64(s.shardNum))
+
+	rrs, err := DoRequests(r.Context(), s.next, reqs)
+	if err != nil {
+		return nil, err
+	}
+
+	// todo: add merging logic here if there are more than one results
+	for _, rr := range rrs {
+		if rr.Response.StatusCode == http.StatusOK {
+			return rr.Response, nil
+		}
+	}
+
+	return nil, fmt.Errorf("trace not found")
+}
+
+// RequestResponse contains a request response and the respective request that was used.
+type RequestResponse struct {
+	Request  *http.Request
+	Response *http.Response
+}
+
+// DoRequests executes a list of requests in parallel. The limits parameters is used to limit parallelism per single request.
+func DoRequests(ctx context.Context, downstream Handler, reqs []*http.Request) ([]RequestResponse, error) {
+	// If one of the requests fail, we want to be able to cancel the rest of them.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Feed all requests to a bounded intermediate channel to limit parallelism.
+	intermediate := make(chan *http.Request)
+	go func() {
+		for _, req := range reqs {
+			intermediate <- req
+		}
+		close(intermediate)
+	}()
+
+	respChan, errChan := make(chan RequestResponse), make(chan error)
+	// todo: make this configurable using limits
+	parallelism := 10
+	if parallelism > len(reqs) {
+		parallelism = len(reqs)
+	}
+	for i := 0; i < parallelism; i++ {
+		go func() {
+			for req := range intermediate {
+				resp, err := downstream.Do(req)
+				if err != nil {
+					errChan <- err
+				} else {
+					respChan <- RequestResponse{req, resp}
+				}
+			}
+		}()
+	}
+
+	resps := make([]RequestResponse, 0, len(reqs))
+	var firstErr error
+	for range reqs {
```

Comment:
It looks like req is not used in this loop.   Would it make sense to use a WaitGroup to know when all resp,err are ready?

Reply:
Yeah, it's just using the len of reqs to pull all of the responses off of `respChan` one at a time.  I think this is fine.